### PR TITLE
server: (router) add stop-timeout option

### LINKED ISSUE #18237: Feature Request: Add watchdog with force-kill timeout for hung child processes in router mode
### Prerequisites

- [x] I am running the latest code. Mention the version if possible as well.
- [x] I carefully followed the [README.md](https://github.com/ggml-org/llama.cpp/blob/master/README.md).
- [x] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggml-org/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

In router mode (--models-dir / --models-preset), when a child server process hangs or becomes unresponsive, the parent router has no mechanism to detect or recover from this state. The monitoring thread blocks indefinitely on subprocess_join() with no timeout.

### Motivation

Production
One bad child shouldn't take down entire router

### Possible Implementation

Proposed Solution (Docker-style watchdog)
Implement a configurable timeout mechanism similar to docker stop

### PR DESCRIPTION
**Pull request by ngxson**
Supersede https://github.com/ggml-org/llama.cpp/pull/18292

Fix https://github.com/ggml-org/llama.cpp/issues/18237

By default, we force-terminate the model instance with a timeout of 10 seconds (same thing on Docker)

### CODE REVIEW

**Comment by ServeurpersoCom** [comment]
Removed previous 18292 from testing server and adding this one

**Comment by ServeurpersoCom** [comment]
After a quick test, no major regressions and the stress test passes without blocking the server like with the previous 18292
But I discovered a edge case problem unrelated to this PR, and I have to redo the test!
Stress test showing a bypass of --models-max 1 (from CLI) because I have more VRAM now and both models can instantiate without OOM!

Sans titre2

Sans titre

Apart from this possible race, I need to hang a child to test the PR!

**Comment by ServeurpersoCom** [comment]
### PR Work: kill -STOP pid
```
(root|~) ps aux | grep llama-server
root        1710  0.0  0.1 14978724 190600 ?     Sl   18:31   0:00 /root/llama.cpp.pascal/build/bin/llama-server --port 8082 --models-max 1 --models-preset backend.ini --webui-config-file frontend.json
root       22271  2.9  1.4 147593944 1439952 ?   SLl  18:41   0:24 /root/llama.cpp.pascal/build/bin/llama-server --host 127.0.0.1 --mlock --port 39839 --sleep-idle-seconds 3600 --webui-config-file frontend.json --alias Dense-Devstral-2-123B-Instruct-2512 --batch-size 128 --ctx-size 131072 --cache-type-k q8_0 --cache-type-v q8_0 --flash-attn on --fit off --kv-unified --model unsloth/Devstral-2-123B-Instruct-2512-GGUF/Devstral-2-123B-Instruct-2512-UD-Q4_K_XL-00001-of-00002.gguf --n-gpu-layers 999 --parallel 4 --ubatch-size 512
root       49158  0.0  0.0   6360  2296 pts/0    S+   18:55   0:00 grep llama-server
(root|~) kill -STOP 22271
```

### And: tail -f llama-server log and unload it from UI OK after timeout
I used stop-timeout = 2 ; Force-kill child process after graceful shutdown timeout in seconds (default: 10)
```
(root|~) tail -f /var/www/ia/models/llama-server.log
srv  log_server_r: request: GET /models/ 127.0.0.1 404
srv  log_server_r: request: POST /v1/models/unload 127.0.0.1 404
srv  log_server_r: request: POST /v1/models/unload 127.0.0.1 404
srv  log_server_r: request: GET /v1/models 127.0.0.1 200
srv  log_server_r: request: GET /props 127.0.0.1 200
srv  log_server_r: request: GET /props 127.0.0.1 200
srv  proxy_reques: proxying request to model Dense-Devstral-2-123B-Instruct-2512 on port 39839
srv  log_server_r: request: GET /props 127.0.0.1 200
srv  log_server_r: request: GET /props 127.0.0.1 200
srv  log_server_r: request: GET /v1/models 127.0.0.1 200
srv        unload: unloading model instance name=Dense-Devstral-2-123B-Instruct-2512
srv    operator(): stopping model instance name=Dense-Devstral-2-123B-Instruct-2512
srv  log_server_r: request: POST /models/unload 127.0.0.1 200
srv  log_server_r: request: GET /v1/models 127.0.0.1 200
srv  log_server_r: request: GET /v1/models 127.0.0.1 200
srv  log_server_r: request: GET /v1/models 127.0.0.1 200
srv  log_server_r: request: GET /v1/models 127.0.0.1 200
srv    operator(): force-killing model instance name=Dense-Devstral-2-123B-Instruct-2512 after 2 seconds timeout
srv    operator(): http client error: Failed to read connection
srv    operator(): http client error: Failed to read connection
srv  log_server_r: request: GET /props 127.0.0.1 500
srv  log_server_r: request: POST /v1/chat/completions 127.0.0.1 500
srv  log_server_r: request: GET /v1/models 127.0.0.1 200
srv  proxy_reques: proxying request to model Dense-Devstral-2-123B-Instruct-2512 on port 39839
srv    operator(): http client error: Could not establish connection
srv  log_server_r: request: GET /props 127.0.0.1 500
srv    operator(): instance name=Dense-Devstral-2-123B-Instruct-2512 exited with status 1
srv  log_server_r: request: GET /v1/models 127.0.0.1 200
^C
(root|~)
```

**Review by ServeurpersoCom — APPROVED** [review]

**Comment by ngxson** [comment]
> Stress test showing a bypass of --models-max 1 (from CLI) because I have more VRAM now and both models can instantiate without OOM!

yeah right, upon hitting `models-max` limit, we current don't wait for the old instance to exit. this will need to be fix

**Comment by ServeurpersoCom** [comment]
I'm testing the latest commit and merging it! Thanks Son

**Comment by ServeurpersoCom** [comment]
My DoS script caused two models to load at the same time, I check the code also

I'm trying to eliminate the race condition by inlining the LRU logic directly in load() while holding the mutex continuously, instead of releasing it between setting LOADING status and counting active models.
and added a lru_unload_in_progress flag to serialize LRU operations and prevent concurrent loads from racing through the count check

This patch work, but it duplicates code from unload_lru() :
```
--- a/tools/server/server-models.cpp
+++ b/tools/server/server-models.cpp
@@ -407,22 +407,57 @@ void server_models::load(const std::string & name) {
     if (!has_model(name)) {
         throw std::runtime_error("model name=" + name + " is not found");
     }
-    unload_lru();

-    std::lock_guard<std::mutex> lk(mutex);
+    std::unique_lock<std::mutex> lk(mutex);
+
+    // Wait if an LRU unload is in progress
+    while (lru_unload_in_progress) {
+        cv.wait(lk);
+    }

-    auto meta = mapping[name].meta;
+    auto & meta = mapping[name].meta;
     if (meta.status != SERVER_MODEL_STATUS_UNLOADED) {
         SRV_INF("model %s is not ready\n", name.c_str());
         return;
     }

-    // prepare new instance info
+    // Set LOADING immediately
+    meta.status = SERVER_MODEL_STATUS_LOADING;
+    meta.last_used = ggml_time_ms();
+
+    // Check if we need to unload LRU (inline with lock held)
+    if (base_params.models_max > 0) {
+        std::string lru_model_name;
+        int64_t lru_last_used = ggml_time_ms();
+        size_t count_active = 0;
+
+        for (const auto & m : mapping) {
+            if (m.second.meta.is_active()) {
+                count_active++;
+                if (m.first != name && m.second.meta.last_used < lru_last_used) {
+                    lru_model_name = m.first;
+                    lru_last_used = m.second.meta.last_used;
+                }
+            }
+        }
+
+        if (!lru_model_name.empty() && count_active > (size_t)base_params.models_max) {
+            SRV_INF("models_max limit reached, removing LRU name=%s\n", lru_model_name.c_str());
+            lru_unload_in_progress = true;  // Block other loads
+            stopping_models.insert(lru_model_name);
+            cv_stop.notify_all();
+            cv.wait(lk, [this, &lru_model_name]() {
+                return mapping[lru_model_name].meta.status == SERVER_MODEL_STATUS_UNLOADED;
+            });
+            lru_unload_in_progress = false;
+            cv.notify_all();  // Wake waiting threads
+        }
+    }
+
+    // Continue normal load (reste du code original)
     instance_t inst;
     inst.meta           = meta;
     inst.meta.port      = get_free_port();
-    inst.meta.status    = SERVER_MODEL_STATUS_LOADING;
-    inst.meta.last_used = ggml_time_ms();

     if (inst.meta.port <= 0) {
         throw std::runtime_error("failed to get a port number");
diff --git a/tools/server/server-models.h b/tools/server/server-models.h
index 7e3353753..3e3f1adf2 100644
--- a/tools/server/server-models.h
+++ b/tools/server/server-models.h
@@ -89,6 +89,8 @@ private:
     std::condition_variable cv_stop;
     std::set<std::string> stopping_models;

+    bool lru_unload_in_progress = false;
+
     common_preset_context ctx_preset;

     common_params base_params;
```

**Comment by ngxson** [comment]
good catch, I think the cleaner fix will be to allow `upload_lru()` to reuse the same lock as `load()`

but now it raises a concern that if we have, for example, models_max=1 and there are 10 incoming requests that each use a different model. now the problem is that 2nd request have to wait for 1st request to finish, then 10th request have to wait for all of the 9 requests in front of it.

I think what's better is to only allow one `upload_lru()` at a time, and throw an error if the second call is invoked.

probably better to create a dedicated issue for this, so I will have a look later.
